urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/stationserver/StationServerImpl.java

```diff
@@ -0,0 +1,741 @@
+	private Properties tcpProperties = new Properties();
+	// this station's udp server to return station's records count;
+	private UDPServer udpServer;
+	private TCPServer tcpServer;
+	// apache logger
+	private Logger log;
+	// all the date's will be interpreted as mm/DD/yyyy
+	private SimpleDateFormat dateFormat = new SimpleDateFormat("mm/DD/yyyy");
+
+	public StationServerImpl startUDPServer(String port) {
+		this.udpServer = new UDPServer(Integer.valueOf(port));
+		this.log.debug(this.stationType.getStationCode()
+				+ ":UDPServer started on port[" + port + "]");
+		return this;
+	}
+
+	/*
+	 * start's UDP server on the port number given in the properites file
+	 */
+	public StationServerImpl startUDPServer() {
+		// read the host and port number to be started on for this station
+		String port = extractHostAndPortNumber((String) udpProperties
+				.get(this.stationType.getStationCode()))[1];
+		return startUDPServer(port);
+	}
+
+	public StationServerImpl startTCPPServer(String port) {
+		try {
+			tcpProperties.setProperty(this.stationType.getStationCode(),
+					InetAddress.getLocalHost().getHostName() + " " + port);
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
+		}
+		this.tcpServer = new TCPServer(Integer.valueOf(port));
+		this.log.debug(this.stationType.getStationCode()
+				+ ":TCPServer started on port[" + port + "]");
+		return this;
+	}
+
+	public void addOtherUDPStationHostNPort(StationType stationType,
+			String host, String port) {
+		udpProperties.setProperty(stationType.getStationCode(), host + " "
+				+ port);
+	}
+
+	public void addOtherTCPStationHostNPort(StationType stationType,
+			String host, String port) {
+		tcpProperties.setProperty(stationType.getStationCode(), host + " "
+				+ port);
+	}
+
+	/*
+	 * start's TCP server on the port number given in the properites file
+	 */
+	public StationServerImpl startTCPPServer() {
+		// read the host and port number to be started on for this station
+		String port = extractHostAndPortNumber((String) tcpProperties
+				.get(this.stationType.getStationCode()))[1];
+		return startTCPPServer(port);
+	}
+
+	/**
+	 * constructor
+	 * 
+	 * @param stationType
+	 */
+	public StationServerImpl(final StationType stationType) {
+		// initialize this station type
+		this.stationType = stationType;
+		// initialize the logger
+		this.log = Logger.getLogger(this.stationType.getStationCode());
+		log.info(stationType.getStationName() + " initialized!");
+		try {
+			// load the udp.properties and tcp.properties file, if not found
+			// write a new file
+			// with the default values
+			udpProperties.load(new FileInputStream("./udp.properties"));
+
+		} catch (FileNotFoundException e) {
+			log.error("config properties not found");
+			applyDefaultUDPSettings();
+		} catch (IOException e) {
+			e.printStackTrace();
+			log.error(e.getMessage());
+		}
+
+		try {
+			tcpProperties.load(new FileInputStream("tcp.properties"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).